File: app/src/main/java/org/schabi/newpipe/database/BasicDAO.java
package org.schabi.newpipe.database;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Update;

import java.util.Collection;
import java.util.List;

import io.reactivex.rxjava3.core.Flowable;

@Dao
public interface BasicDAO<Entity> {
    /* Inserts */
    @Insert
    long insert(Entity entity);

    @Insert
    List<Long> insertAll(Collection<Entity> entities);

    /* Searches */
    Flowable<List<Entity>> getAll();

    Flowable<List<Entity>> listByService(int serviceId);

    /* Deletes */
    @Delete
    void delete(Entity entity);

    int deleteAll();

    /* Updates */
    @Update
    int update(Entity entity);

    @Update
    void update(Collection<Entity> entities);
}

File: app/src/main/java/org/schabi/newpipe/database/BasicDAO.kt
/*
 * SPDX-FileCopyrightText: 2017-2022 NewPipe contributors <https://newpipe.net>
 * SPDX-FileCopyrightText: 2025 NewPipe e.V. <https://newpipe-ev.de>
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package org.schabi.newpipe.database

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import kotlinx.coroutines.flow.Flow

@Dao
interface BasicDAO<Entity> {

    /* Inserts */
    @Insert
    suspend fun insert(entity: Entity): Long

    @Insert
    suspend fun insertAll(entities: Collection<Entity>): List<Long>

    /* Searches */
    fun getAll(): Flow<List<Entity>>

    fun listByService(serviceId: Int): Flow<List<Entity>>

    /* Deletes */
    @Delete
    suspend fun delete(entity: Entity)

    @Query("") // This is a placeholder as BasicDAO is generic, but subclasses override it
    suspend fun deleteAll(): Int

    /* Updates */
    @Update
    suspend fun update(entity: Entity): Int

    @Update
    suspend fun update(entities: Collection<Entity>)
}
